### PR TITLE
Vcmi map format (part1)

### PR DESCRIPTION
**Pull request by alexvins**
_WIP. Nothing to test here yet. Code comments and suggestions are appreciated. General discussion http://forum.vcmi.eu/viewtopic.php?t=1083_

**Goals**
- [x] Load random map in editor
- [x] Load random map in engine as normal map
- [x] Make maps actually playable
- [x] Do not break saves for now
- [ ] Fine tune map format

**Some implementation details**

1. Json serialization is "heavy" so I do not care about virtual calls, but I care about huge template magic.
Anyway is will be integrated with #194 at some 
2. Some things in engine need to be rewritten, for now used workarounds. Next part of this task will contain savegame breakage and CMap and map objects re-factoring

### CODE REVIEW

File: lib/mapObjects/ObjectTemplate.cpp [hunk 1]
```diff
@@ -249,11 +252,97 @@ void ObjectTemplate::readJson(const JsonNode &node)
...
+			for(auto type : allowedTerrains)
+			{
+				JsonNode value(JsonNode::DATA_STRING);
+				value.String() = GameConstants::TERRAIN_NAMES[type.num];
+				data.push_back(value);
+			}
+		}
+	}
+//todo: use lookup?
+	auto tileToChar = [&](const ui8 & tile) -> char
```
Comment (DjWarmonger) [inline lib/mapObjects/ObjectTemplate.cpp]: Doesn't looki like lookup could b any faster, but less readable.

File: lib/mapObjects/ObjectTemplate.cpp [hunk 2]
```diff
@@ -249,11 +252,97 @@ void ObjectTemplate::readJson(const JsonNode &node)
...
+		visitDirs[2].String().resize(3);
+
+		visitDirs[0].String()[0] = (visitDir & 1)	? '+' : '-';
+		visitDirs[0].String()[1] = (visitDir & 2)	? '+' : '-';
+		visitDirs[0].String()[2] = (visitDir & 4)	? '+' : '-';
+		visitDirs[1].String()[2] = (visitDir & 8)	? '+' : '-';
+		visitDirs[2].String()[2] = (visitDir & 16)	? '+' : '-';
+		visitDirs[2].String()[1] = (visitDir & 32)	? '+' : '-';
+		visitDirs[2].String()[0] = (visitDir & 64)	? '+' : '-';
+		visitDirs[1].String()[0] = (visitDir & 128) ? '+' : '-';
```
Comment (DjWarmonger) [inline lib/mapObjects/ObjectTemplate.cpp]: This missing tab won't let me sleep.

File: lib/spells/CSpellHandler.cpp [hunk 3]
```diff
@@ -799,11 +799,12 @@ const std::string CSpellHandler::getTypeName() const
 	return "spell";
 }
 
-CSpell * CSpellHandler::loadFromJson(const JsonNode & json)
+CSpell * CSpellHandler::loadFromJson(const JsonNode & json, const std::string & identifier)
 {
 	using namespace SpellConfig;
 
 	CSpell * spell = new CSpell();
+	spell->identifier = identifier;
```
Comment (DjWarmonger) [inline lib/spells/CSpellHandler.cpp]: Why this is needed?

Comment (alexvins) [inline lib/spells/CSpellHandler.cpp]: Same as for other entities. To be able to convert spell back to string id.

**Comment by DjWarmonger** [comment]
As far as I see, RMG changed look and work well.

**Pull request merged**